src/Brainstorm.hs

```diff
@@ -0,0 +1,106 @@
+
+-- | Transpose a MIDI note by some amount. We could also use intervals here if needed.
+transposeM :: MIDI -> Int -> MIDI
+transposeM = undefined
+
+-- | Pitch in scientific pitch notation (SPN)
+data Pitch = Pitch {getNote :: Note, octave :: Int}
+
+-- | Smart constructor for Pitch
+pitchFrom :: Note -> Int -> Pitch
+pitchFrom = undefined
+
+-- | Convert pitch to a MIDI
+pitchToMIDI :: Pitch -> MIDI
+pitchToMIDI = undefined
+
+-- | Convert pitch to a frequency
+pitchToFreq :: Pitch -> Frequency
+pitchToFreq = midiToFreq . pitchToMIDI
+
+-- | Transpose a pitch by some interval. We could also use integers here if needed.
+transposeP :: Pitch -> Interval -> Pitch
+transposeP = undefined
+
+-- | Convert MIDI to the simplest possible pitches (i.e. using the fewest accidentals)
+midiToPitches :: MIDI -> [Pitch]
+midiToPitches = undefined
+
+-- | A chord with a bass note (not the best name, but hey) that can be specified or not
+data ChordWithBass = ChordWithBass {getChord :: Chord,
+                                    getBass :: Maybe Note}
+
+-- | Checks whether the voicing is in root position (i.e. bass note is chord root)
+isRootPosition :: ChordWithBass -> Bool
+isRootPosition = undefined
+
+-- | A voicing is a chord with a bass note potentially specified, and a list of pitches.
+data Voicing = Voicing {getCWB :: ChordWithBass,
+                        getPitches :: [Pitch]}
+
+-- | Smart constructor for voicings.
+voicingFrom :: ChordWithBass -> [Pitch] -> Maybe Voicing
+voicingFrom = undefined
+
+-- | There will be a number of such functions for different voicing strategies.
+generateVoicing :: ChordWithBass -> [Pitch]
+generateVoicing = undefined
+
+-- | Given two voicings, returns the voice leading distance between them in half steps.
+voiceLeadingDistance :: Voicing -> Voicing -> Int
+voiceLeadingDistance = undefined
+
+-- | Given a voicing and a chord we want to reach, find the voicing with the minimum voice leading distance.
+voiceLeadFromTo :: Voicing -> ChordWithBass -> Voicing
+voiceLeadFromTo = undefined
+
+newtype Progression = Progression {getChords :: [ChordWithBass]}
+
+-- | Given a chord progression and an initial voicing, voice lead each successive chord
+-- to minimize voice leading distance from the previous one.
+progressionToVoicings :: Progression -> Voicing -> [Voicing]
+progressionToVoicings = undefined
+
+-- | Tritone sub an entire progression by subbing each individual chord.
+-- We can come up with a number of different substitutions that operate on
+-- the whole progression rather than just individual chords.
+tritoneSub :: Progression -> Progression
+tritoneSub = undefined
+
+-- | Given a progression, a number of beats per bar, and a list of
+-- durations (in number of beats), i.e. one duration per chord in progression,
+-- we can create a lead sheet datatype.
+data LeadSheet = LeadSheet {getProg :: Progression,
+                            getMeter :: Int,
+                            getDurations :: [Int]}
```

Comment:
Seeing this type makes me _so_ excited; it's elegant, easily expandable if we need more information later, and totally captures all the information we expect to be in the initial language frontend. All to say, this type is _screaming_ "it's almost time for a release!"

On that note: We might want to consider getting this stuff done, and actually preparing for a release soon after ( this prep will take a fair amount of bookkeeping effort; checking documentation consistency, getting a README in order, confirming that public APIs are correct, etc), since we have such a huge library of stuff that's frankly already useful to people who, like us,  are at the intersection of music and CS. I've personally been thinking of this project as both an effort to put together really solid Haskell libraries for representing musical ideas and an effort to provide a convenient tool (ie the language) that builds on those libraries' power. I would totally defer to your judgment on any of this, though; after all, you own the concept as far as I'm concerned :)

Reply:
We do seem so close to a prototype! 

I love the idea of getting a release ready after this -- this would've all just been a random thought without your input making it actually seem/be possible, so if you think we could be ready soon, that means a ton. I defer all decision-making about what needs to be done to be release-ready to you, since you actually know what that takes. I completely agree with the broader idea about this project as both representation of musical ideas and convenient tooling -- I also see this as a first step in structuring/reducing the state space of music production and thereby making feasible the automation of increasingly greater compositional tasks.